presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -16,53 +16,284 @@
+
+        private int compare(WindowNode o1, WindowNode o2)
+        {
+            int comparison = comparePartitionBy(o1, o2);
+            if (comparison != 0) {
+                return comparison;
+            }
+
+            comparison = compareOrderBy(o1, o2);
+            if (comparison != 0) {
+                return comparison;
+            }
+
+            // If PartitionBy and OrderBy clauses are identical, let's establish an arbitrary order to prevent non-deterministic results of swapping WindowNodes in such a case
+            return o1.getId().toString().compareTo(o2.getId().toString());
+        }
+
+        private int comparePartitionBy(WindowNode o1, WindowNode o2)
+        {
+            Iterator<Symbol> iterator1 = o1.getPartitionBy().iterator();
+            Iterator<Symbol> iterator2 = o2.getPartitionBy().iterator();
+
+            while (iterator1.hasNext() && iterator2.hasNext()) {
+                Symbol symbol1 = iterator1.next();
+                Symbol symbol2 = iterator2.next();
+
+                int partitionByComparison = symbol1.compareTo(symbol2);
+                if (partitionByComparison != 0) {
+                    return partitionByComparison;
+                }
+            }
+
+            if (iterator1.hasNext()) {
+                return 1;
+            }
+            if (iterator2.hasNext()) {
+                return -1;
+            }
+            return 0;
+        }
+
+        private int compareOrderBy(WindowNode o1, WindowNode o2)
+        {
+            Iterator<Symbol> iterator1 = o1.getOrderBy().iterator();
+            Iterator<Symbol> iterator2 = o2.getOrderBy().iterator();
+
+            while (iterator1.hasNext() && iterator2.hasNext()) {
+                Symbol symbol1 = iterator1.next();
+                Symbol symbol2 = iterator2.next();
+
+                int orderByComparison = symbol1.compareTo(symbol2);
+                if (orderByComparison != 0) {
+                    return orderByComparison;
+                }
+                else {
+                    int sortOrderComparison = o1.getOrderings().get(symbol1).compareTo(o2.getOrderings().get(symbol2));
+                    if (sortOrderComparison != 0) {
+                        return sortOrderComparison;
+                    }
+                }
+            }
+
+            if (iterator1.hasNext()) {
+                return 1;
+            }
+            if (iterator2.hasNext()) {
+                return -1;
+            }
+            return 0;
+        }
+    }
+
+    /**
+     * Looks for the pattern (ProjectNode*)TargetNode, and rewrites it to TargetNode(ProjectNode*),
+     * returning an empty option if it can't rewrite the projects, for example because they rely on
+     * the output of the TargetNode.
+     * @param projects the nodes above the target, bottom first.
+     */
+    private static <N extends PlanNode> Optional<N> pullUnaryNodeAboveProjects(
+            Class<N> targetClass,
+            N target,
```

Comment:
Why not to accept `WindowNode` here?

Reply:
Yes, the generality is not relevant now that the function isn't in Util any more.  I'll make it specific to WindowNode, and I'll pull it in as a protected static in ManipulateAdjacentWindowsOverProjects.